pkg/imgpkg/imageset/image_set.go

```diff
@@ -77,57 +79,135 @@ func (o *ImageSet) Import(imgOrIndexes []imagedesc.ImageOrIndex,
+	for _, item := range imgOrIndexes {
+		item := item // copy
+
+		go func() {
+			importThrottle.Take()
+			defer importThrottle.Done()
+
+			o.verifyProcessedImages(item, errChVerifyImages, importRepo, registry, importedImages)
 		}()
 	}
 
+	err = checkForAnyAsyncErrors(imgOrIndexes, errChVerifyImages)
+	if err != nil {
+		return nil, err
+	}
+
+	return importedImages, nil
+}
+
+func checkForAnyAsyncErrors(imgOrIndexes []imagedesc.ImageOrIndex, errCh chan error) error {
 	for i := 0; i < len(imgOrIndexes); i++ {
 		err := <-errCh
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
+	return nil
+}
 
-	return importedImages, nil
+func addImageOrImageIndexToTaggableMapForMultiWrite(item imagedesc.ImageOrIndex, errCh chan error, importRepo regname.Repository, registry ImagesReaderWriter, imageOrIndexesToWrite map[regname.Reference]regremote.Taggable, lock *sync.Mutex) {
+	itemDigest, err := item.Digest()
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	tag := fmt.Sprintf("imgpkg-%s-%s", itemDigest.Algorithm, itemDigest.Hex)
+	uploadTagRef, err := regname.NewTag(fmt.Sprintf("%s:%s", importRepo.Name(), tag))
+	if err != nil {
+		errCh <- fmt.Errorf("Building upload tag image ref: %s", err)
+		return
+	}
+
+	switch {
+	case item.Image != nil:
+		itemRef, err := regname.ParseReference(item.Ref())
+		if err != nil {
+			errCh <- fmt.Errorf("Unable to parse reference: %s: %s", item.Ref(), err)
+			return
+		}
+
+		imageToWrite := regv1.Image(*item.Image)
+
+		if imageBlobsCanBeMounted(itemRef, uploadTagRef) {
+			imageToWrite, err = item.MountableImage(registry)
+			if err != nil {
+				imageToWrite = regv1.Image(*item.Image)
+			}
+		}
+
+		lock.Lock()
+		defer lock.Unlock()
+		imageOrIndexesToWrite[uploadTagRef] = imageToWrite
+	case item.Index != nil:
+		lock.Lock()
+		defer lock.Unlock()
+		imageOrIndexesToWrite[uploadTagRef] = *item.Index
+	}
+
+	errCh <- nil
+}
+
+func (o *ImageSet) verifyProcessedImages(item imagedesc.ImageOrIndex, errCh chan error, importRepo regname.Repository, registry ImagesReaderWriter, importedImages *ProcessedImages) {
+	existingRef, err := regname.NewDigest(item.Ref())
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	importDigestRef, err := o.tagAndVerifyImage(item, existingRef, importRepo, registry)
+	if err != nil {
+		errCh <- fmt.Errorf("Importing image %s: %s", existingRef.Name(), err)
+		return
+	}
+
+	var regImage regv1.Image
+	if item.Image != nil {
+		regImage = *item.Image
+	}
+	var regImageIndex regv1.ImageIndex
+	if item.Index != nil {
+		regImageIndex = *item.Index
+	}
+	importedImages.Add(ProcessedImage{
+		UnprocessedImageRef: UnprocessedImageRef{existingRef.Name(), item.Tag()},
+		DigestRef:           importDigestRef.Name(),
+		Image:               regImage,
+		ImageIndex:          regImageIndex,
+	})
+	errCh <- nil
 }
 
-func (o *ImageSet) importImage(item imagedesc.ImageOrIndex,
+func (o *ImageSet) tagAndVerifyImage(item imagedesc.ImageOrIndex,
```

Comment:
since MultiWrite is doing all the work, we no longer have "real time" logging of image imports (o.logger.Write...). this is going to be weird for the user as im guessing it looks like there is a big gap of "silence" in output while things get uploaded. and then we print importing... more or less at once.

Reply:
Yeah this is true. This is what my second comment raised. I outline a couple of ideas on how to handle this.

Reply:
@cppforlife talked, we don't think this PR should be blocked by this concern. It is a valid concern. We also don't think changing the log message "copying.... this may take some time" is sufficient. 

- we think that this should be solved by contributing upstream to https://github.com/google/go-containerregistry/issues/952
-- This will change the UX to a 'progress bar' view.